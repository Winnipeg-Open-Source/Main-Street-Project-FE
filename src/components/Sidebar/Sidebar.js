import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Flex } from 'pcln-design-system';
import Modal from 'components/Modal';
import Link from 'components/Link';

const COLLAPSED_WIDTH = 56;
const EXPANDED_WIDTH = 200;

const LinksWrapper = styled(Flex)`
    & > * {
        padding: 8px 0;
    }
`;

function Sidebar ({ className, isMobileSidebar, isCollapsed, currentPathname, routes, onClose }) {
    const Wrapper = isMobileSidebar ? Modal : React.Fragment;

    const width = isCollapsed && !isMobileSidebar ? COLLAPSED_WIDTH : EXPANDED_WIDTH;
    const backgroundColor = isMobileSidebar ? 'inherit' : 'tertiary';
    const padding = isMobileSidebar ? 0 : 3;

    const modalProps = isMobileSidebar ? {
        className: className,
        isMobile: isMobileSidebar,
        isOpen: !isCollapsed,
        title: 'Navigation',
        onClose,
    } : {};

    return (
        <Wrapper {...modalProps}>
            <LinksWrapper className={className} flexDirection='column' width={width} color={backgroundColor} p={padding}>
                {routes && routes.map(route => (
                    <Link
                        key={route.path}
                        isActive={currentPathname === route.path}
                        to={route.path}
                        icon={route.icon}
                        onClick={isMobileSidebar ? onClose : null}
                    >
                        {route.label}
                    </Link>
                ))}
            </LinksWrapper>
        </Wrapper>
    );
}

Sidebar.displayName = 'Sidebar';

Sidebar.propTypes = {
    className: PropTypes.string,
    isMobileSidebar: PropTypes.bool,
    isCollapsed: PropTypes.bool,
    currentPathname: PropTypes.string,
    routes: PropTypes.array,
    onClose: PropTypes.func,
};

Sidebar.defaultProps = {
    className: '',
};

export default memo(Sidebar);
