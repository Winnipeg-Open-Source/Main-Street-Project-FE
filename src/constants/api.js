export const BASE_API_PATH = 'http://localhost:8080';
export const DONORS_API_PATH = BASE_API_PATH + '/api/donors';
