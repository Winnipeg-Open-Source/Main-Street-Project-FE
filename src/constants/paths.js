export const LANDING_PATH = '/';
export const DONORS_PATH = '/donors';
export const DONOR_PATH = '/donors/:id';
export const PRODUCTS_PATH = '/products';
export const PRODUCT_PATH = '/products/:id';
export const PRODUCTS_CREATE_PATH = '/products/create';
export const PRODUCTS_EDIT_PATH = '/products/edit/:id';
export const DONATIONS_NEW_PATH = '/donations/new';
export const REQUISITION_NEW_PATH = '/requisition/new';

const getPath = (path, id) => {
    return id ? `${path}/${id}` : path;
};

export const getDonorPath = (id) => {
    return getPath(DONORS_PATH, id);
};

export const getProductPath = (id) => {
    return getPath(PRODUCTS_PATH, id);
};

export const getProductCreatePath = (id) => {
    return getPath(PRODUCTS_CREATE_PATH, id);
};

export const getProductEditPath = (id) => {
    return getPath(PRODUCTS_EDIT_PATH, id);
};