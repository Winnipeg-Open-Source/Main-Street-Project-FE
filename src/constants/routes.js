import React from 'react';
import { Document, ExitToApp, Gallery, Home, Loyalty, Map, User } from 'pcln-icons';
import {
    LANDING_PATH,
    DONATIONS_PATH,
    DONORS_PATH,
    ITEMS_PATH,
    LOCATIONS_PATH,
    LOGOUT_PATH,
    REQUISITIONS_PATH,
} from 'constants/paths';

const routes = [
    {
        label: 'Home',
        path: LANDING_PATH,
        icon: <Home />,
    },
    {
        label: 'Donors',
        path: DONORS_PATH,
        icon: <User />,
    },
    {
        label: 'Donations',
        path: DONATIONS_PATH,
        icon: <Loyalty />,
    },
    {
        label: 'Items',
        path: ITEMS_PATH,
        icon: <Gallery />,
    },
    {
        label: 'Locations',
        path: LOCATIONS_PATH,
        icon: <Map />,
    },
    {
        label: 'Requisitions',
        path: REQUISITIONS_PATH,
        icon: <Document />,
    },
    {
        label: 'Logout',
        path: LOGOUT_PATH,
        icon: <ExitToApp />,
    },
];

export default routes;
