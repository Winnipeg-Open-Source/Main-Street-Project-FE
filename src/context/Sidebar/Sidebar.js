import { createContext } from 'react';

export default createContext({
    isCollapsed: false,
    onSidebarToggle: null,
});
