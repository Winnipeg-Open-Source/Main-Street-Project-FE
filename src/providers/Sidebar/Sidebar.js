import React from 'react';
import SidebarContext from 'context/Sidebar';
import useToggle from 'hooks/useToggle';

function Sidebar ({ children }) {
    const [ isCollapsed, onToggle ] = useToggle(true);

    const props = {
        isCollapsed,
        onSidebarToggle: onToggle,
    };

    return (
        <SidebarContext.Provider value={props}>
            {children}
        </SidebarContext.Provider>
    );
}

Sidebar.displayName = 'SidebarProvider';

export default Sidebar;
