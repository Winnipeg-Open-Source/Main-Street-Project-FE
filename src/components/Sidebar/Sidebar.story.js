import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import Sidebar from './Sidebar';
import routes from 'constants/routes';
import { LANDING_PATH } from '../../constants/paths';

export const IsAdmin = () => (
    <BrowserRouter>
        <Sidebar isAdmin currentPathname={LANDING_PATH} routes={routes} />
    </BrowserRouter>
);

export const Expanded = () => (
    <BrowserRouter>
        <Sidebar currentPathname={LANDING_PATH} routes={routes} />
    </BrowserRouter>
);

export const Collapsed = () => (
    <BrowserRouter>
        <Sidebar isCollapsed currentPathname={LANDING_PATH} routes={routes} />
    </BrowserRouter>
);

export default {
    title: 'Sidebar',
    component: Sidebar,
};
