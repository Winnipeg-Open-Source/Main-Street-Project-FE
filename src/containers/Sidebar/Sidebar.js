import React from 'react';
import { useLocation } from 'react-router-dom';
import SidebarComponent from 'components/Sidebar';
import useResponsive from 'hooks/useResponsive';
import useSidebar from 'hooks/context/useSidebar';
import routes from 'constants/routes';

function Sidebar () {
    const { currentBreakpoint } = useResponsive();
    const { isCollapsed, onSidebarToggle } = useSidebar();
    const { pathname } = useLocation();

    const isSmallBreakpoint = currentBreakpoint === 'xs' || currentBreakpoint === 'sm';

    return (
        <SidebarComponent
            isMobileSidebar={isSmallBreakpoint}
            isCollapsed={isCollapsed}
            currentPathname={pathname}
            routes={routes}
            onClose={onSidebarToggle}
        />
    );
}

Sidebar.displayName = 'Sidebar';

export default Sidebar;
