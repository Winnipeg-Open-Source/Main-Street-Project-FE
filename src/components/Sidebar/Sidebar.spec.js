import React from 'react';
import { fireEvent } from '@testing-library/react';
import renderWithThemeAndRouter from 'tests/helpers/renderWithThemeAndRouter';
import Sidebar from './Sidebar';
import routes from 'constants/routes';
import { LANDING_PATH, PRODUCTS_PATH, DONORS_PATH } from '../../constants/paths';

describe ('Sidebar', () => {
    it ('renders without crashing', () => {
        const { asFragment } = renderWithThemeAndRouter(<Sidebar />);

        expect(asFragment()).toMatchSnapshot();
    });

    it ('can navigate to products page', () => {
        const { getByText, history } = renderWithThemeAndRouter(
            <Sidebar currentPathname={LANDING_PATH} routes={routes} />
        );
        const productsButton = getByText('Products');

        expect(history.location.pathname).toBe(LANDING_PATH);
        fireEvent.click(productsButton);
        expect(history.location.pathname).toBe(PRODUCTS_PATH);
    });

    it ('can navigate to donors page', () => {
        const { getByText, history } = renderWithThemeAndRouter(
            <Sidebar currentPathname={LANDING_PATH} routes={routes} />
        );
        const donorsButton = getByText('Donors');

        expect(history.location.pathname).toBe(LANDING_PATH);
        fireEvent.click(donorsButton);
        expect(history.location.pathname).toBe(DONORS_PATH);
    });
});
