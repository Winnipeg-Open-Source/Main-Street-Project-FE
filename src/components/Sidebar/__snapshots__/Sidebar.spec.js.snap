// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-pbJYR hNzbLI sc-pkfPC jpeBiC"
  >
    <div
      class="sc-AxjAm sc-AxiKw sc-pRtcU duboaV"
      color="tertiary.dark"
      width="200"
    />
  </div>
</DocumentFragment>
`;
